pennylane/resources/second_quantization.py

```diff
@@ -15,374 +15,529 @@
-        raise ValueError("br must be a positive integer.")
+        if br <= 0 or not isinstance(br, int):
+            raise ValueError("br must be a positive integer.")
+
+        if alpha <= 0 or not isinstance(alpha, int):
+            raise ValueError("alpha must be a positive integer.")
 
-    if alpha <= 0 or not isinstance(alpha, int):
-        raise ValueError("alpha must be a positive integer.")
+        if beta <= 0 or not isinstance(beta, int):
+            raise ValueError("beta must be a positive integer.")
 
-    if beta <= 0 or not isinstance(beta, int):
-        raise ValueError("beta must be a positive integer.")
+        rank_rm = rank_r * rank_m
+
+        # eta is computed based on step 1.(a) in page 030305-41 of PRX Quantum 2, 030305 (2021)
+        eta = np.array([np.log2(n) for n in range(1, rank_r + 1) if rank_r % n == 0])
+        eta = int(np.max([n for n in eta if n % 1 == 0]))
+
+        nxi = np.ceil(np.log2(rank_max))  # Eq. (C14) of PRX Quantum 2, 030305 (2021)
+        nl = np.ceil(np.log2(rank_r + 1))  # Eq. (C14) of PRX Quantum 2, 030305 (2021)
+        nlxi = np.ceil(np.log2(rank_rm + n / 2))  # Eq. (C15) of PRX Quantum 2, 030305 (2021)
+
+        bp1 = nl + alpha  # Eq. (C27) of PRX Quantum 2, 030305 (2021)
+        bo = nxi + nlxi + br + 1  # Eq. (C29) of PRX Quantum 2, 030305 (2021)
+        bp2 = nxi + alpha + 2  # Eq. (C31) of PRX Quantum 2, 030305 (2021)
+
+        # cost is computed using Eq. (C39) of PRX Quantum 2, 030305 (2021)
+        cost = (
+            9 * nl - 6 * eta + 12 * br + 34 * nxi + 8 * nlxi + 9 * alpha + 3 * n * beta - 6 * n - 43
+        )
+        cost += self._qrom_cost((rank_r, 1, 0, bp1, -1))[0]
+        cost += self._qrom_cost((rank_r, 1, 0, bo, -1))[0]
+        cost += self._qrom_cost((rank_r, 1, 0, 1, 0))[0] * 2
+        cost += self._qrom_cost((rank_rm, n / 2, rank_rm, n * beta, 0))[0]
+        cost += self._qrom_cost((rank_rm, n / 2, rank_rm, 2, 0))[0] * 2
+        cost += self._qrom_cost((rank_rm, n / 2, rank_rm, 2 * bp2, -1))[0]
+
+        return int(cost)
+
+    def gate_cost(self, n, lamb, error, rank_r, rank_m, rank_max, br=7, alpha=10, beta=20):
+        r"""Return the total number of Toffoli gates needed to implement the double factorization
+        algorithm.
+
+        The expression for computing the cost is taken from Eqs. (45) and (C39) of
+        [`PRX Quantum 2, 030305 (2021) <https://journals.aps.org/prxquantum/abstract/10.1103/PRXQuantum.2.030305>`_].
+
+        Args:
+            n (int): number of molecular spin-orbitals
+            lamb (float): 1-norm of a second-quantized Hamiltonian
+            error (float): target error in the algorithm
+            rank_r (int): rank of the first factorization of the two-electron integral tensor
+            rank_m (float): average rank of the second factorization of the two-electron tensor
+            rank_max (float): maximum rank of the second factorization of the two-electron tensor
```

Comment:
Same as before. Either this has to be `int` or its `isinstance` check has to be with `float`.